Math/NumberTheory/Moduli/Sqrt.hs

```diff
@@ -10,33 +10,85 @@
 --
 
 {-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE ViewPatterns #-}
 {-# LANGUAGE CPP          #-}
 
 module Math.NumberTheory.Moduli.Sqrt
-  ( sqrtModP
-  , sqrtModPList
-  , sqrtModP'
-  , tonelliShanks
-  , sqrtModPP
-  , sqrtModPPList
+  ( -- Constructor and @'unPrime'@ are exported in order to do pattern-matching.
+    FieldCharacteristic(..)
+  , unPrime
+  -- Exported for testing purposes. Normally should not be used.
+  , primeF
+  , toFieldCharacteristic
+  , sqrtModList
+  , sqrtModMaybe
+  , sqrtModExact
   , sqrtModF
   , sqrtModFList
   ) where
 
+import Control.Arrow (first)
 import Control.Monad (liftM2)
 import Data.Bits
+import Data.Coerce
 import Data.List (nub)
 import GHC.Integer.GMP.Internals
 
 import Math.NumberTheory.Moduli.Chinese
 import Math.NumberTheory.Moduli.Jacobi
+import Math.NumberTheory.Powers.General (highestPower)
+import Math.NumberTheory.Primes.Types
 import Math.NumberTheory.Primes.Sieve (sieveFrom)
+import Math.NumberTheory.Primes.Testing.Probabilistic (isPrime)
 import Math.NumberTheory.Utils (shiftToOddCount, splitOff)
 
+import Math.NumberTheory.Utils.FromIntegral
+
+-- | Extended flag structure to check if square root is computed in finite field.
+data FieldCharacteristic = FieldCharacteristic (Prime Integer) Int
+  deriving Show
+
+-- | Duplicate functions from @'instance UniqueFactorisation Integer'@. They cannot be imported because of cycle imports.
```

Comment:
`Math.NumberTheory.Primes.Types` is already in scope, so we can use `PrimeNat` / `unPrimeNat` directly.

Reply:
They're not exported anymore since https://github.com/cartazio/arithmoi/pull/108/commits/33f258515cf029d75a994b44480123923f0b8850
So these are just specified wrappers around `fromIntegral` used to coerce `PrimeNat = PrimeNat Natural` with `Integer` in the module. Since we have `sqrt` functions specified for `Integer` it's ok. Relates #100